cellmlmanip/units.py

```diff
@@ -417,24 +553,40 @@ def traverse(self, expr: sympy.Expr):
                                                   self.ureg.dimensionless)
 
                     # magnitude contains an unresolved symbol, we lose it here!
+                    # we don't lose it - the unit will be dimensionless - we are just not able to
+                    # determine the magnitude of the result
                     return 1 * self.ureg.dimensionless
 
                 logger.critical('Exp operand is not dimensionless: %s', expr)
-                return None
+                raise InputArgumentsMustBeDimensionlessError('%s' % expr)
             # trig. function on any dimensionless operand is dimensionless
-            elif str(expr.func) in TRIG_FUNCTIONS and self._is_dimensionless(quantity_per_arg[0]):
-                return 1 * self.ureg.dimensionless
+            elif str(expr.func) in TRIG_FUNCTIONS:
+                if self._is_dimensionless(quantity_per_arg[0]):
+                    return 1 * self.ureg.dimensionless
+                raise InputArgumentsMustBeDimensionlessError('%s' % expr)
 
             # if the function has exactly one dimensionless argument
             if len(quantity_per_arg) == 1 and self._is_dimensionless(quantity_per_arg[0]):
                 # assume the result is dimensionless
                 return 1 * self.ureg.dimensionless
-        elif expr == sympy.pi:
-            return math.pi * self.ureg.dimensionless
+
         elif expr.is_Derivative:
             out = quantity_per_arg[0] / quantity_per_arg[1]
             return out
 
+        # constants in cellml that are all specified as dimensionless
+        elif expr == sympy.pi:
+            return math.pi * self.ureg.dimensionless
+        elif expr == sympy.E:
+            return math.e * self.ureg.dimensionless
+        elif expr == sympy.oo:
```

Comment:
oo and nan need unit tests

Reply:
I have

Reply:
# constants
        assert unit_calculator.traverse(sp.pi).units == ureg.dimensionless
        assert unit_calculator.traverse(sp.E).units == ureg.dimensionless
        assert unit_calculator.traverse(sp.oo).units == ureg.dimensionless
        assert unit_calculator.traverse(sp.nan).units == ureg.dimensionless

Reply:
but coverage claims they are not tested !!!

Reply:
Codecov is correct.
I looked at it and it seems that both
`sp.oo.is_Number == True`
`sp.nan.is_Number == True`

Counter intuitive I know, but it can be fixed easily by moving those 2 checks before the is-number check I think.